README.md
# dynamics-clj

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2018 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-03-26
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-03-26
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/dynamics-clj/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/dynamics-clj/compare/0.1.0...0.1.1


test/dynamics_clj/core_test.clj
(ns dynamics-clj.core-test
  (:require [clojure.test :refer :all]
            [dynamics-clj.core :refer [create-record]]))

(create-record config "contacts" {:firstname "Billy" :lastname "Bob"})

;; (
;;  def config {;; Azure AD Username
;;              :username "user@myorg.onmicrosoft.com"
;;              ;; Azure AD password
;;              :password "secretpassword"
;;              ;; Azure Active Directory > App Registrations > Endpoints > OAuth Token Endpoint
;;              :tokenendpoint "https://login.microsoftonline.com/some-guid-here/oauth2/token"
;;              ;; Dynamics CRM Url
;;              :crmorg "https://myorg.crm.dynamics.com"
;;              ;; You must register a new app in Azure AD > App Registrations. The type must be "Native".
;;              ;; This will generate a new application with an application id (synonym for clientid) 
;;              ;; You must then go to Settings > Required Permissions (scroll right) > click Add > Dynamics CRM Online > click "Grant Permissions" (required!)
;;              :clientid  "some-guid"
;;              ;; Dynamics CRM > Settings > Customizations > Developer Settings > Web API Url (2016)
;;              :crmwebapipath "https://myorg.crm.dynamics.com/api/data/v9.0/"})

;; (def real-config (read-string (slurp "./crm-config.edn")))

;; (def id ())

;; (def multi-response "{\"@odata.context\":\"https://myorg.crm.dynamics.com/api/data/v9.0/$metadata#contacts(fullname)\",\"value\":[{\"fullname\":\"Billy Bob\",\"contactid\":\"95cb1af9-3934-e811-a834-000d3a33b1e4\"}]}")

;; (def single-response "{\"@odata.context\":\"https://d4hines.crm.dynamics.com/api/data/v9.0/$metadata#contacts(fullname)/$entity\",\"fullname\":\"Billy Bob\",\"contactid\":\"e2577e87-3a34-e811-a834-000d3a33b1e4\"}")

;; (def token "secrettoken")



nodejs/dwa.js
var DynamicsWebApi = require('dynamics-web-api');
var AuthenticationContext = require('adal-node').AuthenticationContext;

//the following settings should be taken from Azure for your application
//and stored in app settings file or in global variables

//OAuth Token Endpoint
var authorityUrl = 'https://login.microsoftonline.com/94348c85-35ff-4ed8-aecd-61b85d0f4c17/oauth2/token';
//CRM Organization URL
var resource = 'https://d4hines.crm.dynamics.com';
//Dynamics 365 Client Id when registered in Azure
var clientId = '00000007-0000-0000-c000-000000000000';
var username = 'd4hines@d4hines.onmicrosoft.com';
var password = 'YechiGlihyechu8';

var adalContext = new AuthenticationContext(authorityUrl);

//add a callback as a parameter for your function
function acquireToken(dynamicsWebApiCallback){
    //a callback for adal-node
    function adalCallback(error, token) {
        if (!error){
            //call DynamicsWebApi callback only when a token has been retrieved
            dynamicsWebApiCallback(token);
        }
        else{
            console.log('Token has not been retrieved. Error: ' + error.stack);
        }
    }

    //call a necessary function in adal-node object to get a token
    adalContext.acquireTokenWithUsernamePassword(resource, username, password, clientId, adalCallback);
}

//create DynamicsWebApi object
var dynamicsWebApi = new DynamicsWebApi({ 
    webApiUrl: 'https:/myorg.api.crm.dynamics.com/api/data/v9.0/',
    onTokenRefresh: acquireToken
});

//call any function
dynamicsWebApi.executeUnboundFunction("WhoAmI").then(function (response) {
    console.log('Hello Dynamics 365! My id is: ' + response.UserId);
}).catch(function(error){
    console.log(error.message);
});

nodejs/pure-adal.js
// Dynamics crm
// ApplicationID: 00000007-0000-0000-c000-000000000000

// dynamics-clj-02
// Secret: test-key: 00LzX9xN5wTQMKGhJLapYa+W9pI2Qf8s/0239VFgkJE=
// ApplicationID: db1d57aa-442c-4878-be22-b08ca8f7a97c
// Reply URL: http://localhost
// AD Tenant ID: 94348c85-35ff-4ed8-aecd-61b85d0f4c17
// TYpe: Web API

// test4
// Application id: 020a1b54-a7d4-4732-8efc-01c13adf8c0c
// Reply URL: http://localhost



console.log('Pure ADAL test')
var AuthenticationContext = require('adal-node').AuthenticationContext;
 
var authorityHostUrl = 'https://login.windows.net';
var tenant = 'd4hines.onmicrosoft.com'; // AAD Tenant name.
var authorityUrl = authorityHostUrl + '/' + tenant;
var applicationId = '020a1b54-a7d4-4732-8efc-01c13adf8c0c'; // Application Id of app registered under AAD.
var resource = 'https://d4hines.crm.dynamics.com'; // URI that identifies the resource for which the token is valid.
var username = 'd4hines@d4hines.onmicrosoft.com';
var password = 'YechiGlihyechu8';
var context = new AuthenticationContext(authorityUrl);
 
function callback (err, tokenResponse) {
  if (err) {
    console.log('well that didn\'t work: ' + err.stack);
  } else {
    console.log(JSON.stringify(tokenResponse));
  }
}

context.acquireTokenWithUsernamePassword(resource, username, password, applicationId, callback);


nodejs/pure-http.js
// oauth token endpoint: https://login.microsoftonline.com/94348c85-35ff-4ed8-aecd-61b85d0f4c17/oauth2/token
'use strict';
process.env.https_proxy = "http://127.0.0.1:8888";
process.env.http_proxy = "http://127.0.0.1:8888";
process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
var https = require('https');
console.log('pure http test')
//set these values to retrieve the oauth token
var crmorg = 'https://d4hines.crm.dynamics.com';
var clientid = '020a1b54-a7d4-4732-8efc-01c13adf8c0c';
var username = 'd4hines@d4hines.onmicrosoft.com';
var userpassword = 'yqW4FyhE9TsQcA6O';
var tokenendpoint = 'https://login.microsoftonline.com/94348c85-35ff-4ed8-aecd-61b85d0f4c17/oauth2/token';
 
//set these values to query your crm data
var crmwebapihost = 'd4hines.api.crm.dynamics.com';
var crmwebapipath = '/api/data/v9.0/contacts'; //basic query to select contacts
 
//remove https from tokenendpoint url
tokenendpoint = tokenendpoint.toLowerCase().replace('https://','');
 
//get the authorization endpoint host name
var authhost = tokenendpoint.split('/')[0];

//get the authorization endpoint path
var authpath = '/' + tokenendpoint.split('/').slice(1).join('/');
 
//build the authorization request
//if you want to learn more about how tokens work, see IETF RFC 6749 - https://tools.ietf.org/html/rfc6749
var reqstring = 'client_id='+clientid;
reqstring+='&resource='+encodeURIComponent(crmorg);
reqstring+='&username='+encodeURIComponent(username);
reqstring+='&password='+encodeURIComponent(userpassword);
reqstring+='&grant_type=password';

//set the token request parameters
var tokenrequestoptions = {
    host: authhost,
    path: authpath,
    method: 'POST',
    headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Content-Length': Buffer.byteLength(reqstring)
    }
};
var start = new Date();
console.log(start);
//make the token request
var tokenrequest = https.request(tokenrequestoptions, function(response) {
    //make an array to hold the response parts if we get multiple parts
    var responseparts = [];
    response.setEncoding('utf8');
    response.on('data', function(chunk) {
        //add each response chunk to the responseparts array for later
        responseparts.push(chunk);		
    });
    response.on('end', function(){
        //once we have all the response parts, concatenate the parts into a single string
        var completeresponse = responseparts.join('');
        //console.log('Response: ' + completeresponse);
        console.log('Token response retrieved . . . ');
        
        //parse the response JSON
        var tokenresponse = JSON.parse(completeresponse);
        
        //extract the token
        var token = tokenresponse.access_token;
        console.log(new Date());
        //pass the token to our data retrieval function
        getData(token);
    });
});
tokenrequest.on('error', function(e) {
    console.error(e);
});
 
//post the token request data
tokenrequest.write(reqstring);
 
//close the token request
tokenrequest.end();
 
 
function getData(token){
    //set the web api request headers
    var requestheaders = { 
        'Authorization': 'Bearer ' + token,
        'OData-MaxVersion': '4.0',
        'OData-Version': '4.0',
        'Accept': 'application/json',
        'Content-Type': 'application/json; charset=utf-8',
        'Prefer': 'odata.maxpagesize=500',
        'Prefer': 'odata.include-annotations=OData.Community.Display.V1.FormattedValue'
    };
    
    //set the crm request parameters
    var crmrequestoptions = {
        host: crmwebapihost,
        path: crmwebapipath,
        method: 'GET',
        headers: requestheaders
    };
    
    //make the web api request
    var crmrequest = https.request(crmrequestoptions, function(response) {
        //make an array to hold the response parts if we get multiple parts
        var responseparts = [];
        response.setEncoding('utf8');
        response.on('data', function(chunk) {
            //add each response chunk to the responseparts array for later
            responseparts.push(chunk);		
        });
        response.on('end', function(){
            console.log(new Date());
            //once we have all the response parts, concatenate the parts into a single string
            var completeresponse = responseparts.join('');
            
            //parse the response JSON
            var collection = JSON.parse(completeresponse).value;
            
            //loop through the results and write out the fullname
            collection.forEach(function (row, i) {
                console.log(row);
            });
        });
    });
    crmrequest.on('error', function(e) {
        console.error(e);
    });
    //close the web api request
    crmrequest.end();
}


src/dynamics_clj/core.clj
(ns dynamics-clj.core
  (:require [cheshire.core :refer [generate-string]]
            [clj-http.client :as client]
            [clojure.string :as str]
            [ring.util.codec :refer [url-encode]]
            [slingshot.slingshot :refer [try+]]))

(defn get-token
  "Retrieves an OAuth2 token from Azure Active Directory.
  Uses username and password credentials.
  `config` is a map of the configuration needed for OAuth with the CRM. See README for examples."
  [config]
  (let [{:keys [crmorg clientid username password tokenendpoint]} config
        reqstring (str "client_id=" clientid
                       "&resource=" (url-encode crmorg)
                       "&username=" (url-encode username)
                       "&password=" (url-encode password)
                       "&grant_type=password")
        options {:body reqstring
                 :headers {"Content-Type" "application/x-www-form-urlencoded"}
                 :as :json}]
    (get-in (client/post tokenendpoint options) [:body :access_token])))

(def crm-options
  "Defaults for clj-http requests, along with the headers needed for the CRM OData requests."
  {:as :json-string-keys
   :headers {"OData-MaxVersion" "4.0",
             "OData-Version" "4.0",
             "Accept" "application/json",
             "Content-Type" "application/json; charset=utf-8",
             "Prefer" "odata.maxpagesize=500,odata.include-annotations=OData.Community.Display.V1.FormattedValue"}})

(defn build-select [fields] (let [field-count (count fields)]
                              (cond (=  0 field-count) nil
                                    (= 1 field-count) (str "$select=" (first fields)))
                              :else (str "$select=" (str/join "," fields))))

(defn retrieve* [config endpoint] (client/get
                                   (str (:crmwebapipath config) endpoint)
                                   (assoc crm-options :oauth-token (get-token config))))

(defn retrieve
  "Retrieves a single entity from CRM.
   `config` is a map of the configuration needed for OAuth with the CRM. See README for examples.
  `entity-col` is the entity's logical collection name (the plural, lowercase).
  `id` is the entity's GUID.
  `fields` is a coll. of logical field names. Pass `nil` to include all fields.
  e.g.
  (retrieve \"contacts\" \"914b2297-bf2f-e811-a833-000d3a33b3a3\" [\"fullname\",\"createdon\"])"
  [config entity-col id fields]
   (try+ (:body (retrieve* config (str entity-col "(" id ")"
                                    (if fields (str "?" (build-select fields)) nil))))
        (catch [:status 404] [] {:status 404 :message (str "Id " id " not found in " entity-col)})))

(defn retrieve-multiple
  "Retrieves a single entity from CRM.
  `config` is a map of the configuration needed for OAuth with the CRM. See README for examples.
  `entity-col` is the entity's logical collection name (the plural, lowercase).
  `fields` is an optional coll. of logical field names. Omit if not required.
  `filter` is an optional OData filter string. Pass `nil` to include all fields.
  e.g
  (retrieve-multiple \"contacts\" [\"fullname\",\"createdon\"] \"firstname eq 'bob'\" )"
  [config entity-col fields filter-expr]
  (let [field-str (if fields (build-select fields) nil)
        filter-str (if filter-expr (str "$filter=" filter-expr) nil)
        combined (filter identity [field-str filter-str])
        final-uri (str entity-col
                       (if (> (count combined) 0) (str "?" (str/join "&" combined)) nil))]
    (get-in (retrieve* config final-uri) [:body "value"])))

(defn create-record
  "Creates a record in CRM, returning the new id.
   `config` is a map of the configuration needed for OAuth with the CRM. See README for examples.
  `entity-col` is the entity's logical collection name (the plural, lowercase).
  `new-record` is a map of the fields to be created on the new entity record
  e.g
  (create-record \"contacts\" {:firstname \"test\" :lastname \"person\"})"
  [config entity-col new-record]
  (let [response (client/post (str (:crmwebapipath config) entity-col)
                              (assoc crm-options :content-type :json
                                     :oauth-token (get-token config)
                                     :body (generate-string new-record)))
        entity-url (get-in response [:headers "OData-EntityId"])]
    (second (str/split entity-url #"\(|\)"))))

(defn update-record
  "Updates an existing CRM record. Returns a ring response map.
  `config` is a map of the configuration needed for OAuth with the CRM. See README for examples.
  `entity-col` is the entity's logical collection name (the plural, lowercase).
  `update-record` is a map of the fields/values to be updated on the new entity record.
  e.g
  (update-record \"contact\" {:firstname \"bob_updated\" :lastname \"person_updated\"})"
  [config entity-col id updated-record]
  (client/patch (str (:crmwebapipath config) entity-col "(" id ")")
                (assoc crm-options :content-type :json
                       :oauth-token (get-token config)
                       :body (generate-string updated-record))))

(defn delete-record
  "Deletes a CRM record. Returns a ring response map.
  `entity-col` is the entity's logical collection name (the plural, lowercase).
  `id` is the entity's GUID.
  e.g
  (delete-record \"contact\" \"9695bd5c-2635-e811-a834-000d3a33b1e4\")"
  [config entity-col id]
  (client/delete (str (:crmwebapipath config) entity-col "(" id ")")
                 (assoc crm-options :content-type :json
                        :oauth-token (get-token config))))

;; EXAMPLES
(comment 
  (def config (read-string (slurp "/crm-config.edn")))

  ;; Retrieve a contact, then get the ID of the first one returned.
  (def id (get (first (retrieve-multiple config "contacts" ["fullname"] "firstname eq 'test'" ))
               "contactid"))

  ;; Retrieve all fields for a contact
  (retrieve config "contacts" id nil)

  ;; Use retrieve* to target arbitrary endpoints. This is useful for metadata.
  (retrieve* config  "EntityDefinitions(286a3bcb-d539-e811-a833-000d3a33bdbd)?$select=LogicalName&$expand=Attributes($select=LogicalName)"))
